contracts/src/v0.8/automation/dev/v2_3/AutomationRegistryBase2_3.sol

```diff
@@ -975,29 +979,43 @@ abstract contract AutomationRegistryBase2_3 is ConfirmedOwner {
 
     PaymentReceipt memory receipt = _calculatePaymentAmount(hotVars, paymentParams);
 
-    uint96 balance = upkeep.balance;
-    uint96 payment = receipt.gasCharge + receipt.premium;
+    // this balance uses the token's decimal places
+    uint96 balanceXDecimals = upkeep.balance;
+    // adjust to 18 decimal places to be comparable with the following payment
+    uint96 balance18Decimals = SafeCast.toUint96(
+      balanceXDecimals * 10 ** (18 - paymentParams.billingTokenParams.decimals)
+    );
+
+    // payment always has 18 decimals
+    uint96 payment18Decimals = receipt.gasCharge + receipt.premium;
 
     // this shouldn't happen, but in rare edge cases, we charge the full balance in case the user
     // can't cover the amount owed
-    if (balance < receipt.gasCharge) {
+    if (balance18Decimals < receipt.gasCharge) {
       // if the user can't cover the gas fee, then direct all of the payment to the transmitter and distribute no premium to the DON
-      payment = balance;
+      payment18Decimals = balance18Decimals;
+      // TODO to confirm: both priceUSD and linkUSD use 18 decimals, so here is safe?
```

Comment:
I think we might get a cleaner solution if we try to handle the decimals problem in `_calculatePaymentAmount()` instead of `_handlePayment()` (see other comment about `PaymentReceipt`). I think the `_handlePayment()` code can remain as-is if we just tweak the `PaymentReceipt` struct so that it always uses the same decimals as the billing token :)